### ISSUE TITLE
reprex_render with new_session does not find input path

### ISSUE THREAD

**Issue by mpadge** [tidyverse/reprex#373]
New `reprex` 2.0.0 does not work on my system. Here's a reprex running through v1:

``` r
devtools::load_all ("/<local>/<path>/<to>/<reprex>", export_all = TRUE)
#> ℹ Loading reprex
packageVersion ("reprex")
#> [1] '2.0.0.9000'
gert::git_commit_info()$id
#> [1] "d6b12e446bcafe104f3ef762ef0d562148b7f072"

# step through reprex_impl & reprex_render_impl fns:
reprex_files <- plan_files(
                           infile = NULL,
                           wd = NULL,
                           outfile = "DEPRECATED")

r_file <- r_file(reprex_files$filebase)

src <- c ("#' ---",
          "#' output: reprex::reprex_document",
          "#' ---",
          "",
          "print(\"this does not work\")")
write_lines(src, r_file)

# do render stuff:
local_rprofile <- path(path_dir(path_real(r_file)), ".Rprofile")
# this works:
reprex_file <- reprex_render_impl(
                                  r_file,
                                  new_session = FALSE,
                                  html_preview = FALSE)
# this does not work:
#reprex_file <- reprex_render_impl(
#                                  r_file,
#                                  new_session = TRUE,
#                                  html_preview = FALSE)

# reprex_render.R:
input <- r_file

# code for new_session == TRUE:
withr::with_dir(
  path_dir(input),
  out <- tryCatch(
    callr::r(
      function(input) {
        rmarkdown::render(
          input,
          quiet = TRUE,
          envir = globalenv(),
          encoding = "UTF-8"
        )
      },
      args = list(input = path_file(input)),
      spinner = FALSE,
      stdout = NULL,
      stderr = NULL
    ),
    error = function(e) e
  )
)
print (out)
#> <callr_status_error: callr subprocess failed: The file 'pious-rhino_reprex.R' does not exist.>
#>  in process 
#> -->
#> <callr_remote_error in abs_path(input):
#>  The file 'pious-rhino_reprex.R' does not exist.>
#>  in process 9864
```

And that leads to v2 being unusable for me, with the failed `withr` call leading straight to the `!isTRUE(std_out_err)` condition and the `abort` call. The following modification fixes:

``` r
devtools::load_all ("/<local>/<path>/<to>/<reprex>", export_all = TRUE)
packageVersion ("reprex")
#> [1] '2.0.0.9000'
gert::git_commit_info()$id
#> [1] "d6b12e446bcafe104f3ef762ef0d562148b7f072"

reprex_files <- plan_files(
                           infile = NULL,
                           wd = NULL,
                           outfile = "DEPRECATED")

r_file <- r_file(reprex_files$filebase)

src <- c ("#' ---",
          "#' output: reprex::reprex_document",
          "#' ---",
          "",
          "print(\"this does not work\")")
write_lines(src, r_file)

# do render stuff:
local_rprofile <- path(path_dir(path_real(r_file)), ".Rprofile")
# this works:
reprex_file <- reprex_render_impl(
                                  r_file,
                                  new_session = FALSE,
                                  html_preview = FALSE)
# this does not work:
#reprex_file <- reprex_render_impl(
#                                  r_file,
#                                  new_session = TRUE,
#                                  html_preview = FALSE)

# reprex_render.R:
input <- r_file

# new_session options:
withr::with_dir(
  path_dir(input),
  out <- tryCatch(
    callr::r(
      function(input) {
        rmarkdown::render(
          input,
          quiet = TRUE,
          envir = globalenv(),
          encoding = "UTF-8"
        )
      },
      #args = list(input = path_file(input)), # does not work
      args = list(input = input), # WORKS
      spinner = FALSE,
      stdout = NULL,
      stderr = NULL
    ),
    error = function(e) e
  )
)
print (out)
#> [1] "/tmp/RtmpkUcLN0/reprex-2b676844f527-godly-vixen/godly-vixen_reprex.md"
```

<details>
<summary>
sessionInfo()
</summary>
<p>
#> R version 4.0.5 (2021-03-31)<br>
#> Platform: x86_64-pc-linux-gnu (64-bit)<br>
#> Running under: Arch Linux<br>
#> <br>
#> Matrix products: default<br>
#> BLAS:   /usr/lib/libblas.so.3.9.1<br>
#> LAPACK: /usr/lib/liblapack.so.3.9.1<br>
#> <br>
#> locale:<br>
#>  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              <br>
#>  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    <br>
#>  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   <br>
#>  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 <br>
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            <br>
#> [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       <br>
#> <br>
#> attached base packages:<br>
#> [1] stats     graphics  grDevices utils     datasets  methods   base     <br>
#> <br>
#> other attached packages:<br>
#> [1] reprex_2.0.0.9000 testthat_3.0.2    nvimcom_0.9-88   <br>
#> <br>
#> loaded via a namespace (and not attached):<br>
#>  [1] pillar_1.5.1      compiler_4.0.5    highr_0.8         remotes_2.3.0    <br>
#>  [5] prettyunits_1.1.1 tools_4.0.5       digest_0.6.27     pkgbuild_1.2.0   <br>
#>  [9] pkgload_1.2.1     evaluate_0.14     memoise_2.0.0     lifecycle_1.0.0  <br>
#> [13] tibble_3.1.0      debugme_1.1.0     pkgconfig_2.0.3   rlang_0.4.10     <br>
#> [17] rstudioapi_0.13   cli_2.4.0         yaml_2.2.1        xfun_0.22        <br>
#> [21] fastmap_1.1.0     withr_2.4.1       styler_1.4.1      stringr_1.4.0    <br>
#> [25] knitr_1.31        desc_1.3.0        fs_1.5.0          vctrs_0.3.7      <br>
#> [29] devtools_2.4.0    rprojroot_2.0.2   glue_1.4.2        R6_2.5.0         <br>
#> [33] processx_3.5.1    fansi_0.4.2       rmarkdown_2.7     sessioninfo_1.1.1<br>
#> [37] clipr_0.7.1       purrr_0.3.4       callr_3.6.0       magrittr_2.0.1   <br>
#> [41] usethis_2.0.1     backports_1.2.1   ps_1.6.0          ellipsis_0.3.1   <br>
#> [45] htmltools_0.5.1.1 utf8_1.2.1        stringi_1.5.3     cachem_1.0.4     <br>
#> [49] crayon_1.4.1<br>
</p>
</details>

<sup>Created on 2021-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

The problematic line is 
https://github.com/tidyverse/reprex/blob/d6b12e446bcafe104f3ef762ef0d562148b7f072/R/reprex_render.R#L104

**Comment by jennybc** [comment]
Is there any way to show this with a simple call to `reprex()`? And, I guess, an input file? I'm finding the above hard to follow.

**Comment by mpadge** [comment]
Yeah, it's tricky because in the state that causes the problem (v2), reprex is broken, and the `rlang::abort` call prevents reprex itself being used to report on the problem. I'll try on a different machine and see if i can derive something more useful.

**Comment by jennybc** [comment]
If you can just give me the set up, i.e. the exact `reprex()` call that you make, I can debug it from there.

**Comment by mpadge** [comment]
I fear that is not going to help here. This is direct output from either `reprex()` with anything in clipboard, or something like `reprex(input=c("some","input"))`:

----

This reprex appears to crash R.
See standard output and standard error for more details.

#### Standard output and error

``` sh
Error in abs_path(input) : The file 'evil-coney_reprex.R' does not exist.
In addition: Warning message:
In normalizePath(path, winslash = winslash, mustWork = mustWork) :
  path[1]="evil-coney_reprex.R": No such file or directory
```

---

Here's a `reprex` using my modified version which replaces Line#104 of `reprex_render.R` with just `input = input` instead of `input = path_file(input)`. This `reprex` demonstrates the core of the problem, which arises because `callr:r` does not inherit the directory specified in `withr::with_dir`.

``` r
setwd ("/<local>/<path>/<to>/reprex")
devtools::load_all (".", export_all = TRUE)
#> ℹ Loading reprex
library (withr)
#> 
#> Attaching package: 'withr'
#> The following objects are masked from 'package:reprex':
#> 
#>     local_options, with_options
input <- tempfile (fileext = ".txt")
write_lines ("something", input)
print (input)
#> [1] "/tmp/RtmpXOdUSw/file1abe2d43ace1.txt"
f <- with_dir (path_dir (input), list.files ())
print (f)
#> [1] "file1abe2689f361.so"  "file1abe2d43ace1.txt"
with_dir (path_dir (input),
          out <- callr::r (function () {
                               list.files ()
      }, args = list ()))
print (out)
#>  [1] "aaa.R"                  "ensure.R"               "highlight.R"           
#>  [4] "prex.R"                 "reprex_document.R"      "reprex_impl.R"         
#>  [7] "reprex_render.R"        "reprex-addin.R"         "reprex-locale.R"       
#> [10] "reprex-options.R"       "reprex-package.R"       "reprex-undo.R"         
#> [13] "reprex.R"               "stringify_expression.R" "sysdata.rda"           
#> [16] "utils-clipboard.R"      "utils-interactivity.R"  "utils-io.R"            
#> [19] "utils-ui.R"             "utils.R"                "venues.R"
```

<sup>Created on 2021-04-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

The `reprex` call was made within the `./R` directory of `reprex`, hence that file list. This is not being called from within RStudio. Here is the same `reprex` called from within RStudio:

``` r
setwd ("/<local>/<path>/<to>/reprex")
devtools::load_all (".", export_all = TRUE)
#> ℹ Loading reprex
library (withr)
#> 
#> Attaching package: 'withr'
#> The following objects are masked from 'package:reprex':
#> 
#>     local_options, with_options
input <- tempfile (fileext = ".txt")
write_lines ("something", input)
print (input)
#> [1] "/tmp/RtmpOqhhyV/file1d2f652b4d31.txt"
f <- with_dir (path_dir (input), list.files ())
print (f)
#> [1] "file1d2f39aa2466.so"  "file1d2f652b4d31.txt"
with_dir (path_dir (input),
          out <- callr::r (function () {
                               list.files ()
      }, args = list ()))
print (out)
#>  [1] "callr-client--55274dd.so" "callr-env-1d2f23ad165"   
#>  [3] "callr-fun-1d2f546348f2"   "callr-scr-1d2f690512b5"  
#>  [5] "callr-sev-1d2f1be30d4b"   "callr-spr-1d2f4a058936"  
#>  [7] "callr-uev-1d2ffe2352a"    "callr-upr-1d2f43cc3d7d"  
#>  [9] "file1d2f39aa2466.so"      "file1d2f652b4d31.txt"
```

<sup>Created on 2021-04-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

So RStudio is doing something to enable `calrr` to inherit the directory parameter, but there is nothing within `callr` to suggest that such inheritance should be expected, the way i understand it? Outside RStudio on my machine, the assumption that the `callr` process inherits the working directory is certainly not applicable. @gaborcsardi?

**Comment by gaborcsardi** [comment]
Can you please install reprex and not run it from `load_all()`? 

And then show a simple `reprex::reprex()` call, and use `clipr::read_clip()` to show the reprex output on the clipboard, if there is any. Like the example below?

> Outside RStudio on my machine, the assumption that the callr process inherits the working directory is certainly not applicable. @gaborcsardi?

Whose assumption is that? `callr::r` certainly does not change the working directory:
```r
❯ R -q -e 'getwd() == callr::r(function() getwd())'
> getwd() == callr::r(function() getwd())
[1] TRUE
```

OTOH reprex does change into a temporary directory:
```r
❯ reprex::reprex(getwd())
ℹ Rendering reprex...
✔ Rendered reprex is on the clipboard.

❯ clipr::read_clip()
[1] "``` r"
[2] "getwd()"
[3] "#> [1] \"/private/var/folders/59/0gkmw1yj2w7bf2dfc3jznv5w0000gn/T/RtmplXJPnL/reprexc2894bf824ac\""
[4] "```"
[5] ""
[6] "<sup>Created on 2021-04-12 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>"
```

**Comment by mpadge** [comment]
Run outside RStudio:

``` r
getwd () == callr::r (function () getwd ())
#> [1] FALSE
```

<sup>Created on 2021-04-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

And within RStudio:

``` r
getwd () == callr::r (function () getwd ())
#> [1] TRUE
```

<sup>Created on 2021-04-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

**Comment by gaborcsardi** [comment]
Maybe you have a local `.Rprofile` that does something? Can you show the actual values of `getwd()` and `callr::r(function() getwd())`?

**Comment by jennybc** [comment]
It feels like we've arrived at a place where reprex has possibly been eliminated and it has something to do with callr or, rather, callr and one specific machine / ?filepath? of @mpadge.

The whole reason for 4b9d578e2e1480ab2b46edab8816c6c6d56a2336 was that @gaborcsardi alerted me to the need to implement the desired working directory before calling `callr::r()`, which I then confirmed for myself as well. And the related test passes on all the platforms covered by CI.

**Comment by mpadge** [comment]
Yeah, it's definitely not a `reprex` issue. That said, @gaborcsardi we still need to resolve this somehow:

``` r
getwd ()
#> [1] "/tmp/RtmpmUm3Ak/reprex-5a41ef71b8b-pure-tayra"
f <- "~/.Rprofile"
ifelse (file.exists (f), readLines (f), "no .Rprofile found")
#> [1] "no .Rprofile found"
callr::r (function () {
              f <- "~/.Rprofile"
              ifelse (file.exists (f),
                      paste0 (".Rprofile found:", readLines (f)),
                      paste0 ("No .Rprofile found; wd=", getwd ()))
          })
#> [1] "No .Rprofile found; wd=/data/code/forks/reprex"
```

<sup>Created on 2021-04-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

Shall I open an issue on `callr`?

**Comment by gaborcsardi** [comment]
> Shall I open an issue on callr?

Sure, please include the output of these, without using reprex, just copy and paste it from the console:

```r
getwd()
normalizePath(getwd())
callr::r(function() getwd())
callr::r(function() normalizePath(getwd()))
```

**Comment by mpadge** [comment]
No, actually it might better be in `withr`?


``` r
getwd()
#> [1] "/data/code/forks/reprex"
normalizePath(getwd())
#> [1] "/data/code/forks/reprex"
callr::r(function() getwd())
#> [1] "/data/code/forks/reprex"
callr::r(function() normalizePath(getwd()))
#> [1] "/data/code/forks/reprex"

withr::with_dir (tempdir(), {
                     c (getwd(),
                        normalizePath(getwd()),
                        callr::r(function() getwd()),
                        callr::r(function() normalizePath(getwd())))
          })
#> [1] "/tmp/RtmpLaG51T"              "/tmp/RtmpLaG51T"             
#> [3] "/data/code/forks/reprex" "/data/code/forks/reprex"
```

**Comment by gaborcsardi** [comment]
I still can't reproduce this, but I think it is probably a race condition that we can work around in processx or callr: https://github.com/r-lib/processx/issues/300
